libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/TelemetryLocationAndProgressDispatcherImpl.kt

```diff
@@ -0,0 +1,151 @@
+package com.mapbox.navigation.core.telemetry
+
+import android.location.Location
+import android.util.Log
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import com.mapbox.navigation.core.telemetry.MapboxNavigationTelemetry.TAG
+import com.mapbox.navigation.core.telemetry.NewRoute.ExternalRoute
+import com.mapbox.navigation.core.telemetry.NewRoute.RerouteRoute
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import java.util.Collections.synchronizedList
+
+internal class TelemetryLocationAndProgressDispatcherImpl(
+    private val scope: CoroutineScope
+) :
+    TelemetryLocationAndProgressDispatcher {
+
+    companion object {
+        private const val LOCATION_BUFFER_MAX_SIZE = 20
+    }
+
+    private val locationsBuffer = synchronizedList(mutableListOf<Location>())
+    private val eventsLocationsBuffer = synchronizedList(mutableListOf<EventLocations>())
+
+    override val routeProgressChannel = Channel<RouteProgress>(Channel.CONFLATED)
+    override val newRouteChannel = Channel<NewRoute>(Channel.CONFLATED)
+
+    override val lastLocation: Location?
+        get() = locationsBuffer.lastOrNull()
+    override var firstLocation: Location? = null
+    override var routeProgress: RouteProgress? = null
+    override var originalRoute = CompletableDeferred<DirectionsRoute>()
+    private val mutex = Mutex()
```

Comment:
Do we really need this `Mutex`?

Reply:
yes. we remove items  with `iterator` and it should be synchronized. fun is `suspend`, so `Mutex` is used for synchronization.

```
private suspend fun accumulatePostEventLocation(location: Location) {
        mutex.withLock {
            val iterator = eventsLocationsBuffer.iterator()
            while (iterator.hasNext()) {
                iterator.next().let {
                    it.addPostEventLocation(location)
                    if (it.postEventLocationsSize() >= LOCATION_BUFFER_MAX_SIZE) {
                        it.onBufferFull()
                        iterator.remove()
                    }
                }
            }
        }
    }
```